engine.py

```diff
@@ -463,14 +463,30 @@ def show_panel(self, panel_id, title, bundle, widget_class, *args, **kwargs):
             widget_instance.setParent(self._get_dialog_parent())
             widget_instance.setObjectName(panel_id)
 
-            dock_widget = QtGui.QDockWidget(title, parent=main_window)
+            # The widget contained in a QDockWidget must be closed in order for app/engine restart to work correctly.
+            # To accomplish this we need to use our own subclass of QDockWidget
+            class DockWidget(QtGui.QDockWidget):
```

Comment:
Could you move this class at the top of the file instead of nesting it here please?

Reply:
I tried this and ran into an issue that I'm looking into. I need to understand the Qt import a bit better I think. When I move this line out of the engine and at the top of the file

`from sgtk.platform.qt import QtCore, QtGui`

I get the following error:

> ERROR: Shotgun: Could not start engine: Failed to load plugin F:\Shotgun\source_code\tk-3dsmax\engine.py. The following error was reported:
> Exception: <type 'exceptions.ImportError'> - cannot import name QtCore
> Traceback (most recent call last):
>   File "C:\Users\Ehsan\AppData\Roaming\Shotgun\ehsan\p68c16.basic.desktop\cfg\install\core\python\tank\util\loader.py", line 59, in load_plugin
>     module = imp.load_source(module_uid, plugin_file)
> 
>   File "F:\Shotgun\source_code\tk-3dsmax\engine.py", line 19, in <module>
>     from sgtk.platform.qt import QtCore, QtGui
> 
> "Shotgun Warning: You need to close the current window dialog before using any more commands."

Reply:
Oh bollocks, that makes total sense. Leave it as is.

The sgtk.platform.qt modules gets initialized only once an engine has been started. If you put this code at the top of the class, it will be execute while the engine code is loaded, so before the engine is started. Just leave it as a nest class then. We've done it elsewhere for apps, I had simply forgotten that for engines we could not do it.

I'm sorry for wasting your time!